android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt

```diff
@@ -113,16 +199,79 @@ abstract class NewWalletDialog2 : TaskLauncherDialog<String>() {
         val name = arguments!!.getString("name")!!
         val password = arguments!!.getString("password")!!
         onCreateWallet(name, password)
-        daemonModel.loadWallet(name, password)
+
+        val multisig = arguments!!.getBoolean("multisig")
+
+        /**
+         * For multisig wallets, wait until all cosigners have been added,
+         * and then create and load the multisig wallet.
+         *
+         * Otherwise, load the created wallet.
+         */
+        if (multisig) {
+            val currentCosigner = arguments!!.getInt("i_signer")
+            val numCosigners = arguments!!.getInt("cosigners")
+            val numSignatures = arguments!!.getInt("signatures")
+
+            // Close the previous cosigner's keystore dialog.
+            // TODO: Consider a better solution for closing dialogs?
+            (targetFragment as KeystoreDialog).dismiss()
```

Comment:
UI objects shouldn't be accessed in doInBackground, because if the screen rotates, a whole new set of objects is created and your actions on the old ones will be lost. Do this in onPostExecute instead, which is guaranteed to have access to the current UI. See the comments on these methods in TaskDialog.

This also applies to `closeDialogs(targetFragment!!)` below.

Reply:
I'd like some of your input on this point.

Right now, every cosigner from 2nd onwards is shown a `KeystoreDialog`, where you choose either the key or the seed (the first cosigner is shown a `MasterPublicKeyDialog` first). And in `onPostExecute()`, the dialog for the next cosigner is called via `showDialog()`. However, if `dismiss()` is moved to `onPostExecute()`, right before the code which creates a new dialog, the old dialog doesn't dismiss in time for `showDialog()` to work. I looked at the debugger, and `fm.findFragmentByTag(tag)` in `Util.kt` returns the `KeystoreDialog` fragment as not null, so it does not create a new one.

I thought about solving it in this way: if we're dealing with the 1st cosigner, show them a new `MasterPublicKeyDialog` as usal. If we're on the 2nd cosigner onward, just change the title of the already existing `KeystoreDialog` to "(currentCosigner of numCosigners)", without closing and reopening it (and that would be the only change between keystore dialogs, anyway).

Would this be alright, or do you have some other idea or tip that would help me out here? Maybe I'm missing something.

Reply:
Basically, this is what I was trying to do:

```kotlin
if (currentCosigner < numCosigners) {
    // The first cosigner sees their master public key; others are prompted for data
    val nextDialog: DialogFragment = if (currentCosigner == 1) {
        MasterPublicKeyDialog()
    } else {
        KeystoreDialog()
    }

    (targetFragment as DialogFragment).dismiss() // dismiss the previous dialog
    
    arguments!!.putInt("i_signer", currentCosigner + 1)

    nextDialog.setArguments(arguments)
    showDialog(this, nextDialog)
```

But, with this, `showDialog()` doesn't work, as it the `KeystoreDialog` fragment still exists when `showDialog()` is triggered.

Reply:
> If we're on the 2nd cosigner onward, just change the title of the already existing KeystoreDialog to "(currentCosigner of numCosigners)", without closing and reopening it (and that would be the only change between keystore dialogs, anyway).

It would be the only visible change, but the arguments would be different because they would contain an additional cosigner. If you can update the arguments of the existing dialog then that would be fine. 

If that doesnt work, another option might be for the KeystoreDialog to close itself rather than remaining open under the NewWalletDialog2. That would prevent you from going back to it, but your ability to go back during this process is already quite limited, so I don't think it makes much of a difference.